src/shogun/multiclass/KNN.cpp

```diff
@@ -123,33 +124,53 @@ SGMatrix<index_t> KNN::nearest_neighbors()
 	distance->precompute_lhs();
 	distance->precompute_rhs();
 
-	//for each test example
-	for (auto i : SG_PROGRESS(range(n)))
+	switch (knn_solver)
 	{
-		COMPUTATION_CONTROLLERS
-		//lhs idx 0..num train examples-1 (i.e., all train examples) and rhs idx i
-		distances_lhs(dists,0,m_train_labels.vlen-1,i);
+	case KNN_BRUTE:
+	{
+		//for each test example
+		for (auto i : SG_PROGRESS(range(n)))
+		{
+			COMPUTATION_CONTROLLERS
+			//lhs idx 0..num train examples-1 (i.e., all train examples) and rhs idx i
+			distances_lhs(dists,0,m_train_labels.vlen-1,i);
 
-		//fill in an array with 0..num train examples-1
-		for (int32_t j=0; j<m_train_labels.vlen; j++)
-			train_idxs[j]=j;
+			//fill in an array with 0..num train examples-1
+			for (int32_t j=0; j<m_train_labels.vlen; j++)
+				train_idxs[j]=j;
 
-		//sort the distance vector between test example i and all train examples
-		Math::qsort_index(dists.vector, train_idxs.vector, m_train_labels.vlen);
+			//sort the distance vector between test example i and all train examples
+			Math::qsort_index(dists.vector, train_idxs.vector, m_train_labels.vlen);
 
 #ifdef DEBUG_KNN
-		io::print("\nQuick sort query {}\n", i);
-		for (int32_t j=0; j<m_k; j++)
-			io::print("{} ", train_idxs[j]);
-		io::print("\n");
+			io::print("\nQuick sort query {}\n", i);
+			for (int32_t j=0; j<m_k; j++)
+				io::print("{} ", train_idxs[j]);
+			io::print("\n");
 #endif
 
-		//fill in the output the indices of the nearest neighbors
-		for (int32_t j=0; j<m_k; j++)
-			NN(j,i) = train_idxs[j];
+			//fill in the output the indices of the nearest neighbors
+			for (int32_t j=0; j<m_k; j++)
```

Comment:
no loop necessary here. `sg_memcpy` could do the job for example, or others as well

Reply:
Okay, I'll change this.

Reply:
How should I get a pointer to the ith row of the matrix for the first parameter of `sg_memcpy()`? Will something like `(NN.matrix+m_k*sizeof(index_t))` work?

Reply:
Or should I use the `SGMatrix<T>::set_column()` function?

Reply:
that would do it as well!